src/skadi/sample.cc

```diff
@@ -80,103 +71,347 @@ uint64_t file_size(const std::string& file_name) {
+    data.map(path, size, POSIX_MADV_SEQUENTIAL);
+    return true;
   }
-  mapped_cache.resize(TILE_COUNT);
 
-  // check the directory for files that look like what we need
-  auto files = get_files(data_source);
-  for (const auto& f : files) {
-    // make sure its a valid index
-    format_t format = format_t::UNKNOWN;
-    auto index = is_hgt(f, format);
-    if (index < mapped_cache.size() && format != format_t::UNKNOWN) {
-      auto size = file_size(f);
-      if (format == format_t::RAW && size != HGT_BYTES) {
-        LOG_WARN("Corrupt elevation data: " + f);
-        continue;
+  inline const char* get_data() const {
+    return data.get();
+  }
+
+  inline format_t get_format() const {
+    return format;
+  }
+
+  inline int& get_usages() {
+    return usages;
+  }
+
+  inline const char* get_unpacked() {
+    return unpacked;
+  }
+
+  inline const char* detach_unpacked() {
+    auto rv = unpacked;
+    unpacked = nullptr;
+    return rv;
+  }
+
+  bool unpack(const char* unpacked) {
+    this->unpacked = unpacked;
+    // for setting where to read compressed data from
+    auto src_func = [this](z_stream& s) -> void {
+      s.next_in = static_cast<Byte*>(static_cast<void*>(data.get()));
+      s.avail_in = static_cast<unsigned int>(data.size());
+    };
+
+    // for setting where to write the uncompressed data to
+    auto dst_func = [this](z_stream& s) -> int {
+      s.next_out = (Byte*)(this->unpacked);
+      s.avail_out = HGT_BYTES;
+      return Z_FINISH; // we know the output will hold all the input
+    };
+
+    // we have to unzip it
+    if (!baldr::inflate(src_func, dst_func)) {
+      LOG_WARN("Corrupt compressed elevation data");
+      format = format_t::UNKNOWN;
+      return false;
+    }
+    return true;
+  }
+
+  static boost::optional<std::pair<uint16_t, format_t>> parse_hgt_name(const std::string& name) {
+    std::smatch m;
+    std::regex e(".*/([NS])([0-9]{2})([WE])([0-9]{3})\\.hgt(\\.gz)?$");
+    if (std::regex_search(name, m, e)) {
+      // enum class format_t{ UNKNOWN = 0, GZIP = 1, RAW = 3 };
+      format_t fmt =
+          m[5].length() ? (m[5] == ".gz" ? format_t::GZIP : format_t::UNKNOWN) : format_t::RAW;
+      auto lon = std::stoi(m[4]) * (m[3] == "E" ? 1 : -1) + 180;
+      auto lat = std::stoi(m[2]) * (m[1] == "N" ? 1 : -1) + 90;
+      if (lon >= 0 && lon < 360 && lat >= 0 && lat < 180) {
+        return std::make_pair(uint16_t(lat * 360 + lon), fmt);
       }
-      mapped_cache[index].first = format;
-      mapped_cache[index].second.map(f, size, POSIX_MADV_SEQUENTIAL);
     }
+    return boost::none;
   }
-}
+};
+
+class tile_data;
+
+struct cache_t {
+  // Cached tiles
+  std::vector<cache_item_t> cache;
+  // Set of reusable tile indexes
+  std::unordered_set<uint16_t> reusable;
+  // Map of pending tiles. No matter how many requests received, only one inflate job per tile
+  // started.
+  std::unordered_map<uint16_t, std::shared_future<tile_data>> pending_tiles;
```

Comment:
Flagging here that this fails to compile in newest gcc and clang:

Gcc
```
/usr/include/c++/11.1.0/bits/unordered_map.h:105:18:   required from ‘class std::unordered_map<short unsigned int, std::shared_future<valhalla::skadi::tile_data> >’
/home/anders/Documents/Programming/mapbox/valhalla/src/skadi/sample.cc:173:63:   required from here
/usr/include/c++/11.1.0/future:906:21: error: static assertion failed: result type must be destructible
  906 |       static_assert(is_destructible<_Res>{},
      |                     ^~~~~~~~~~~~~~~~~~~~~~~
/usr/include/c++/11.1.0/future:906:21: note: ‘std::is_destructible<valhalla::skadi::tile_data>()’ evaluates to false
/usr/include/c++/11.1.0/future: In instantiation of ‘class std::future<valhalla::skadi::tile_data>’:
/home/anders/Documents/Programming/mapbox/valhalla/src/skadi/sample.cc:329:23:   required from here
/usr/include/c++/11.1.0/future:769:21: error: static assertion failed: result type must be destructible
  769 |       static_assert(is_destructible<_Res>{},
      |                     ^~~~~~~~~~~~~~~~~~~~~~~
/usr/include/c++/11.1.0/future:769:21: note: ‘std::is_destructible<valhalla::skadi::tile_data>()’ evaluates to false
/usr/include/c++/11.1.0/future: In instantiation of ‘class std::promise<valhalla::skadi::tile_data>’:
/home/anders/Documents/Programming/mapbox/valhalla/src/skadi/sample.cc:342:27:   required from here
/usr/include/c++/11.1.0/future:1065:21: error: static assertion failed: result type must be destructible
 1065 |       static_assert(is_destructible<_Res>{},
      |                     ^~~~~~~~~~~~~~~~~~~~~~~
/usr/include/c++/11.1.0/future:1065:21: note: ‘std::is_destructible<valhalla::skadi::tile_data>()’ evaluates to false
[106/180] Building CXX object src/thor/CMakeFiles/valhalla-thor.dir/astar_bss.cc.o


# version
g++ -v
Using built-in specs.
COLLECT_GCC=g++
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/lto-wrapper
Target: x86_64-pc-linux-gnu
Configured with: /build/gcc/src/gcc/configure --prefix=/usr --libdir=/usr/lib --libexecdir=/usr/lib --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=https://bugs.archlinux.org/ --enable-languages=c,c++,ada,fortran,go,lto,objc,obj-c++,d --with-isl --with-linker-hash-style=gnu --with-system-zlib --enable-__cxa_atexit --enable-cet=auto --enable-checking=release --enable-clocale=gnu --enable-default-pie --enable-default-ssp --enable-gnu-indirect-function --enable-gnu-unique-object --enable-install-libiberty --enable-linker-build-id --enable-lto --enable-multilib --enable-plugin --enable-shared --enable-threads=posix --disable-libssp --disable-libstdcxx-pch --disable-libunwind-exceptions --disable-werror gdc_include_dir=/usr/include/dlang/gdc
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 11.1.0 (GCC) 
```

Clang
```
/home/anders/Documents/Programming/mapbox/valhalla/src/skadi/sample.cc:173:63: note: in instantiation of template class 'std::unordered
_map<unsigned short, std::shared_future<valhalla::skadi::tile_data>>' requested here                                                   
  std::unordered_map<uint16_t, std::shared_future<tile_data>> pending_tiles;                                                           
                                                              ^                                                                        In file included from /home/anders/Documents/Programming/mapbox/valhalla/src/skadi/sample.cc:6:                                        /usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../include/c++/11.1.0/future:769:21: error: cannot initialize object parameter of type 'const std::integral_constant<bool, false>' with an expression of type 'is_destructible<valhalla::skadi::tile_data>'               static_assert(is_destructible<_Res>{},                                                                                                               ^~~~~~~~~~~~~~~~~~~~~~~                                                                                            
/home/anders/Documents/Programming/mapbox/valhalla/src/skadi/sample.cc:329:23: note: in instantiation of template class 'std::future<va
lhalla::skadi::tile_data>' requested here                                                                                              
    auto future = it->second;                                                                                                          
                      ^                                                                                                                
In file included from /home/anders/Documents/Programming/mapbox/valhalla/src/skadi/sample.cc:6:
/usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../include/c++/11.1.0/future:1065:21: error: cannot initialize object paramet
er of type 'const std::integral_constant<bool, false>' with an expression of type 'is_destructible<valhalla::skadi::tile_data>'
      static_assert(is_destructible<_Res>{},
                    ^~~~~~~~~~~~~~~~~~~~~~~
/home/anders/Documents/Programming/mapbox/valhalla/src/skadi/sample.cc:342:27: note: in instantiation of template class 'std::promise<v
alhalla::skadi::tile_data>' requested here
  std::promise<tile_data> promise;
                          ^
3 errors generated.

# Version

clang++ --version
clang version 12.0.1
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
```

Reply:
It depends on STL version on your machine. It builds fine with libc++ 9, but won't build on libc++ 11.1 :(

Reply:
Oh, seems like I didn’t build master since the merge.. can smth be done about that? If I can’t build Valhalla anymore on arch linux, I’d have a semi-serious problem

Reply:
Ohh another arch user! I have a semi-serious problem too :)

Reply:
lets make an issue out of it and who ever gets to it first can take a stab at it

Reply:
https://github.com/valhalla/valhalla/issues/3316

Reply:
gcc with libstdc++ 11.1.0 throws this error if `tile_data` forward declared before cache_t  
https://godbolt.org/z/8sYhf8Wqb
while clang 12.0.1 with libc++ works fine. `-stdlib=libc++` 
https://godbolt.org/z/vPYTdb3aW